include/seastar/core/shared_mutex.hh

```diff
@@ -234,6 +237,44 @@ with_lock(shared_mutex& sm, Func&& func) noexcept {
     }
 }
 
+/// \brief Construct a RAII-based shared lock corresponding to a given \ref shared_mutex object.
+///
+/// Since constructors cannot be exectued asynchronously, use this function
+/// to shared lock a \ref shared_mutex and construct an std::shared_lock object
+/// that will unlock the mutex when it's being destroyed.
+///
+/// The caller is responsible for keeping the corresponding \ref shared_mutex object
+/// alive at least until the returned lock is destroyed.
+inline future<std::shared_lock<shared_mutex>> get_shared_lock(shared_mutex& mutex) {
```

Comment:
We should consider making the mutex type a template parameter so it could be extended/overridden, even for testing.